src/standalone/import-export/importTracker.node.ts

```diff
@@ -162,77 +166,91 @@ export class ImportTracker implements IExtensionSingleActivationService, IDispos
         } catch (ex) {
             // Can fail on CI, if the notebook has been closed or the like
             if (!isCI) {
                 throw ex;
             }
         }
-
-        this.lookForImports(result);
     }
 
-    @capturePerfTelemetry(EventName.HASHED_PACKAGE_PERF)
-    private checkDocument(document: TextDocument) {
-        this.pendingChecks.delete(document.fileName);
-        const lines = this.getDocumentLines(document);
-        this.lookForImports(lines);
-    }
-
-    private async sendTelemetry(packageName: string) {
-        // No need to send duplicate telemetry or waste CPU cycles on an unneeded hash.
-        if (this.sentMatches.has(packageName)) {
-            return;
-        }
-        this.sentMatches.add(packageName);
-        // Hash the package name so that we will never accidentally see a
-        // user's private package name.
-        const hash = await getTelemetrySafeHashedString(packageName);
-        sendTelemetryEvent(EventName.HASHED_PACKAGE_NAME, undefined, { hashedNamev2: hash });
+    private async checkDocument(
+        document: TextDocument,
+        resourceType: ResourceTypeTelemetryProperty['resourceType'],
+        when: 'onExecution' | 'onOpenCloseOrSave'
+    ) {
+        await this.sendTelemetryForImports(this.getDocumentLines(document), resourceType, when);
     }
 
-    private lookForImports(lines: (string | undefined)[]) {
+    @capturePerfTelemetry(EventName.HASHED_PACKAGE_PERF)
+    private lookForImports(lines: string[]) {
+        const packageNames: string[] = [];
         try {
             for (const s of lines) {
+                // No need of regex if we don't have imports
+                if (!s.includes('import') && !s.includes('from')) {
+                    continue;
+                }
                 const match = s ? ImportRegEx.exec(s) : null;
                 if (match !== null && match.groups !== undefined) {
                     if (match.groups.fromImport !== undefined) {
                         // `from pkg ...`
-                        this.sendTelemetry(match.groups.fromImport).ignoreErrors();
+                        packageNames.push(match.groups.fromImport);
                     } else if (match.groups.importImport !== undefined) {
                         // `import pkg1, pkg2, ...`
-                        const packageNames = match.groups.importImport
-                            .split(',')
-                            .map((rawPackageName) => rawPackageName.trim());
-                        // Can't pass in `this.sendTelemetry` directly as that rebinds `this`.
-                        packageNames.forEach((p) => this.sendTelemetry(p));
+                        packageNames.push(
+                            ...match.groups.importImport.split(',').map((rawPackageName) => rawPackageName.trim())
+                        );
                     }
                 }
             }
-        } catch {
+        } catch (ex) {
```

Comment:
I'm surprised typescript didn't want this as _ex for being unused...wonder why?

Reply:
damn, i added that to test something.. we don't need it